### ISSUE TITLE
[QUESTION]: Can strings be used in a kernel?

### ISSUE THREAD

**Issue by El-Gor-do** [m4rs-mt/ILGPU#1298]
### Question

I have many years experience in writing high performance C# code but am only just starting to learn about writing GPU code so please forgive my ignorance about the possibilities and capabilities of GPU.

I have some calculations that currently run on CPU and I'm investigating how to port them to GPU. Some calculations involve string concatenation and slicing substrings. I will need to copy string values from host to device in order to perform the calculations on device - I've got this working for `int`/`float`/`double` types but I haven't figured out how to copy `string`s.

`string` is a reference type so it can't be used as a type in `Allocate1D<T>`. I've considered converting `string` to a UTF8 byte array but `byte[]` is also a reference type so it still can't be used in `Allocate1D<T>`. Here is what I've tried:
```
Accelerator device = ...
int arrayLength = 10;
var IntValues = device.Allocate1D<int>(arrayLength);
var StringValues = device.Allocate1D<string>(arrayLength);  // compile error - string is not a non-nullable value type
var ByteArrayValues = device.Allocate1D<byte[]>(arrayLength);  // compile error - byte[] is not a non-nullable value type
```

Is it actually possible to use strings with ILGPU?

### Environment

- ILGPU version: 1.5.1
- .NET version: .NET 8
- Operating system: Windows 11
- Hardware (if GPU-related): NVIDIA GeForce GTX 3060

**Comment by amabilee** [comment]
`ILGPU` is designed primarily for numeric and low-level data processing on the GPU, and it doesn’t natively support complex data types like C# `string` or `byte[]` because these are reference types. 

I think you can convert each string to a **flat byte[]** (using a fixed encoding, such as UTF8), which is a contiguous block of memory that represents the string data.

**Comment by El-Gor-do** [comment]
I can convert `string` to `byte[]` using `System.Text.Encoding.UTF8.GetBytes(string)` then `Marshal.AllocHGlobal(byteArrayLength)` to get pointers to the byte arrays but how do I copy the byte array contents to the device?

**Comment by amabilee** [comment]
`var deviceBytes = accelerator.Allocate1D<byte>(totalLength)`
allocates a contiguous block of memory on the GPU for the byte array.

`deviceBytes.CopyFromHost(concatenatedBytes)`
copies the data from the host byte array concatenatedBytes to the device memory allocated for deviceBytes.

**Comment by El-Gor-do** [comment]
Sorry, I think I phrased my question correctly. I know how to copy a single `string` to `byte[]` and then `Allocate1D<byte>(length)`, what I'm trying to do is copy a `string[]` to the device. Each string can have a different length. The kernel will perform some operations on these strings, e.g. `output[index] = SomeFunction(input[index])` and each `output[index]` element may have a different length than its corresponding `input[index]` element. In this case, is it possible to copy some representation of the `string[]` to the device?

Maybe `Allocate2DDenseY<byte>(new LongIndex2D(elementCount, maxStringLength + 1))`? Each string is treated as a fixed length, null-terminated string up to `maxStringLength` characters long. But then I have to allocate `maxStringLength + 1` bytes per string even if the string value is shorter, and the kernel can't output strings longer than `maxStringLength`.

**Comment by MoFtZ** [comment]
hi @El-Gor-do. As you have noticed, there is no support for strings in ILGPU. You would have to manage everyone yourself, such as treating the string as a sequence of bytes.

You might be interested in [RAPIDS libcubf](https://developer.nvidia.com/blog/mastering-string-transformations-in-rapids-libcudf/), which is a C++ library provided by Nvidia. There is no integration with ILGPU, but it might give you some ideas.

**Comment by El-Gor-do** [comment]
I managed to pass string[] to the device as ArrayView2D<byte> but after playing a bit with my implementation, I realised that it wasn't going to be feasible to implement the string operations that I need. The most immediate problem was dealing with case-insensitive string comparisons without being able to use `StringComparer.OrdinalIgnoreCase`. I suspect it's not going to be worth the effort to get my string operations running on GPU.
